google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionImpl.java

```diff
@@ -195,6 +198,43 @@ public CommitResponse writeAtLeastOnceWithOptions(
     }
   }
 
+  @Override
+  public ServerStream<BatchWriteResponse> batchWriteAtLeastOnce(
+      Iterable<MutationGroup> mutationGroups) throws SpannerException {
+    return batchWriteAtLeastOnceWithOptions(mutationGroups);
+  }
+
+  @Override
+  public ServerStream<BatchWriteResponse> batchWriteAtLeastOnceWithOptions(
+      Iterable<MutationGroup> mutationGroups, TransactionOption... transactionOptions)
+      throws SpannerException {
+    setActive(null);
+    Options batchWriteRequestOptions = Options.fromTransactionOptions(transactionOptions);
+    List<BatchWriteRequest.MutationGroup> mutationGroupsProto =
+        MutationGroup.toListProto(mutationGroups);
+    final BatchWriteRequest.Builder requestBuilder =
+        BatchWriteRequest.newBuilder().setSession(name).addAllMutationGroups(mutationGroupsProto);
+    if (batchWriteRequestOptions.hasPriority() || batchWriteRequestOptions.hasTag()) {
+      RequestOptions.Builder requestOptionsBuilder = RequestOptions.newBuilder();
+      if (batchWriteRequestOptions.hasPriority()) {
+        requestOptionsBuilder.setPriority(batchWriteRequestOptions.priority());
+      }
+      if (batchWriteRequestOptions.hasTag()) {
+        requestOptionsBuilder.setTransactionTag(batchWriteRequestOptions.tag());
+      }
+      requestBuilder.setRequestOptions(requestOptionsBuilder.build());
+    }
+    Span span = tracer.spanBuilder(SpannerImpl.BATCH_WRITE).startSpan();
+    try (Scope s = tracer.withSpan(span)) {
+      return spanner.getRpc().batchWriteAtLeastOnce(requestBuilder.build(), this.options);
+    } catch (Throwable e) {
+      TraceUtil.setWithFailure(span, e);
+      throw SpannerExceptionFactory.newSpannerException(e);
+    } finally {
+      span.end(TraceUtil.END_SPAN_OPTIONS);
+    }
+  }
+
```

Comment:
This method seems to be very similar to `writeAtLeastOnceWithOptions`. Could we combine the two so they share more of the code?

Reply:
Moved the `RequestOptions` building logic to a separate private method.